cdc/processor/processor.go

```diff
@@ -0,0 +1,758 @@
+				ResolvedTs:   checkpointTs,
+			}, nil
+		}
+		return position, nil
+	})
+	return true
+}
+
+// lazyInitImpl create Filter, SchemaStorage, Mounter instances at the first tick.
+func (p *processor) lazyInitImpl(ctx context.Context) error {
+	if !p.firstTick {
+		return nil
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	p.cancel = cancel
+	ctx = util.PutChangefeedIDInCtx(ctx, p.changefeed.ID)
+
+	errCh := make(chan error, 16)
+	go func() {
+		// there are some other objects need errCh, such as sink and sink manager
+		// but we can't ensure that all the producer of errCh are non-blocking
+		// It's very tricky that create a goroutine to receive the local errCh
+		// TODO(leoppro): we should using `pkg/context.Context` instead of standard context and handle error by `pkg/context.Context.Throw`
+		for {
+			select {
+			case <-ctx.Done():
+				close(errCh)
+				return
+			case err := <-errCh:
+				if err == nil {
+					return
+				}
+				p.sendError(err)
+			}
+		}
+	}()
+
+	var err error
+	p.filter, err = filter.NewFilter(p.changefeed.Info.Config)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	p.schemaStorage, err = p.createAndDriveSchemaStorage(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	p.mounter = entry.NewMounter(p.schemaStorage, p.changefeed.Info.Config.Mounter.WorkerNum, p.changefeed.Info.Config.EnableOldValue)
+	go func() {
+		p.sendError(p.mounter.Run(ctx))
+	}()
+
+	opts := make(map[string]string, len(p.changefeed.Info.Opts)+2)
+	for k, v := range p.changefeed.Info.Opts {
+		opts[k] = v
+	}
+
+	// TODO(neil) find a better way to let sink know cyclic is enabled.
+	if p.changefeed.Info.Config.Cyclic.IsEnabled() {
+		cyclicCfg, err := p.changefeed.Info.Config.Cyclic.Marshal()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		opts[mark.OptCyclicConfig] = string(cyclicCfg)
+	}
+	opts[sink.OptChangefeedID] = p.changefeed.ID
+	opts[sink.OptCaptureAddr] = p.captureInfo.AdvertiseAddr
+	s, err := sink.NewSink(ctx, p.changefeed.ID, p.changefeed.Info.SinkURI, p.filter, p.changefeed.Info.Config, opts, errCh)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	checkpointTs := p.changefeed.Info.GetCheckpointTs(p.changefeed.Status)
+	p.sinkManager = sink.NewManager(ctx, s, errCh, checkpointTs)
+
+	// Clean up possible residual error states
+	p.changefeed.PatchTaskPosition(func(position *model.TaskPosition) (*model.TaskPosition, error) {
+		if position != nil && position.Error != nil {
+			position.Error = nil
+		}
+		return position, nil
+	})
+
+	log.Info("run processor",
+		zap.String("capture-id", p.captureInfo.ID), util.ZapFieldCapture(ctx),
+		zap.String("changefeed-id", p.changefeed.ID))
+	return nil
+}
+
+// handleErrorCh listen the error channel and throw the error if it is not expected.
+func (p *processor) handleErrorCh(ctx context.Context) error {
+	var err error
+	select {
+	case err = <-p.errCh:
+	default:
+		return nil
+	}
+	p.cancel()
+	if util.WaitTimeout(&p.wg, waitTableCloseTimeout) {
```

Comment:
Why don't wait until wait group is done, if timeout there could be some routine leak?

Reply:
when the `MySQLSinkHangLongTime` failpoint is enabled, table pipeline will not exit

https://github.com/pingcap/ticdc/blob/master/cdc/sink/mysql.go#L832